### ISSUE TITLE
Refresh Token Expiration

### ISSUE THREAD

**Issue by yurtesen** [dmfs/oauth2-essentials#51]
Refresh tokens also have expiration times. But it does not seem to be possible to access with this library?

**Comment by dmfs** [comment]
I can't remember seeing anything about refresh token expiration times. Where did you see that?
Usually refresh tokens are valid until the server issues a new one during a token refresh request. If the server chooses to invalidate a refresh token (or the user revokes access of an app) your client will receive an error and you'll have to go through the authentication flow again.

**Comment by yurtesen** [comment]
For example popular FOSOauthServerBundle has expiration in [documentation](https://github.com/FriendsOfSymfony/FOSOAuthServerBundle/blob/master/Resources/doc/configuration_reference.md) Default seems to be 14 days.
```
            # Changing tokens and authcode lifetime
            #access_token_lifetime: 3600
            #refresh_token_lifetime: 1209600
            #auth_code_lifetime: 30
```
I also see the `expires_at` field in my database for FOSOauthServerBundle. But also I found at least this [old article](https://blog.cloud-elements.com/oauth-2-0-access-refresh-token-guide) about refresh tokens expiring in some servers.

In either case, actually I now tested fosoauthserverbundle and even though it has expiration for refresh tokens, the json output does not include it...doh... So I guess I should make a bug report to them :)

Thanks for the response!

**Comment by dmfs** [comment]
It's correct that, with some servers, refresh tokens expire. Yet, there is no expiration date in the server response and RFC 6749 doesn't specify any. Note that there is no dedicated request to refresh the refresh token. Instead a server will just issue a new refresh token when you refresh the access token. So you should always check the result of `OAuth2AccessToken.hasRefreshToken()` when you refresh the access token and update your refresh token if this returns `true`. Most servers with expiring refresh tokens will return a new refresh token with every token refresh grant.

If your client didn't refresh the access token before the expiration date of the refresh token, you'll most likely have to re-authorize the user. Given the expiration times of 6 Months or 60 days I wouldn't consider this an issue. If your client didn't use an API for this amount of time, the user probably doesn't use it any more.

The refresh token expiration mechanism is certainly meant to automatically revoke access of abandoned clients.

**Comment by dmfs** [comment]
I'm closing this issue because RFC 6749 doesn't specify a refresh token expiration date. This is essentially an internal server implementation detail.

**Comment by yurtesen** [comment]
Thanks for the information. Your Oauth2 library is actually very neat and useful. It solved a lot of problems for me. I blame Oauth2 for being inconsistent. I started to hate Oauth2 already :) Some features seem like they were made to make things more complicated only. I am surprised they did not have unspecified amount of refresh tokens for "security", like 2ndRefreshToken for refreshing the 1stRefreshToken. 

Just as a note, I tested with FOSOauthServerBundle by setting refresh token expiration to 10 seconds. It removes/disables the refresh token when it expires and one can't use it anymore. So in such cases user must login/authenticate using different method (eg. username/password) again. Thats perfectly fine by me :)
